python/pyspark/sql/connect/session.py

```diff
@@ -289,65 +295,48 @@ def createDataFrame(
         else:
             _data = list(data)
 
-            if _schema is None and isinstance(_data[0], (Row, dict)):
-                if isinstance(_data[0], dict):
-                    # Sort the data to respect inferred schema.
-                    # For dictionaries, we sort the schema in alphabetical order.
-                    _data = [dict(sorted(d.items())) for d in _data]
+            if isinstance(_data[0], dict):
+                # Sort the data to respect inferred schema.
+                # For dictionaries, we sort the schema in alphabetical order.
+                _data = [dict(sorted(d.items())) for d in _data]
+
+            elif not isinstance(_data[0], (Row, tuple, list, dict)):
+                # input data can be [1, 2, 3]
+                # we need to convert it to [[1], [2], [3]] to be able to infer schema.
+                _data = [[d] for d in _data]
 
+            try:
                 _inferred_schema = self._inferSchemaFromList(_data, _cols)
-                if _cols is not None:
-                    for i, name in enumerate(_cols):
-                        _inferred_schema.fields[i].name = name
-                        _inferred_schema.names[i] = name
+            except Exception:
+                # For cases like createDataFrame([("Alice", None, 80.1)], schema)
+                # we can not infer the schema from the data itself.
+                warnings.warn("failed to infer the schema from data")
```

Comment:
there will be an edge case that can not be supported in Connect:

`createDataFrame([(None, None, None)],  'ddl schema')`

since we don't have a ddl-parser in Python Client.